orttraining/orttraining/python/orttraining_pybind_state.cc

```diff
@@ -928,6 +938,15 @@ for every transfered tensor.
         [](const std::string& key, const std::unordered_set<size_t> edges) -> void {
           GradientDefinitionRegistry::Instance().SetStopGradientEdgesForNode(key, edges);
         });
+
+#if defined(ENABLE_TRAINING) && defined(ENABLE_TRAINING_ON_DEVICE)
+  m.def("save_checkpoint",
```

Comment:
This the initial implementation actually, I was then hesitating about 1). is loading the user ONNX file in python offline code required? 2). passing those protoes through python bindings involved unnecessary copies. 

So want to confirm with you, given a user ONNX model, the motivation we wanted to load model in Python offline code, is to mark requires_grad, am I right? If so, is that enough we just pass the model path and trainable weight names into our python bind api?

Reply:
The assumption is that the user only has an inference only onnx model. The offline step helps the user with:
1. Compose a suitable loss node to the inference only graph.
2. Build a gradient graph for the user implicitly.
3. Build the optimizer graph for the user.
4. Allow users to set metadata for the graph parameters (like requires_grad and potentially parameter groupings for optimizer in the future).
5. Prepare the final training graph in a format that can be directly consumed by the online training step (move parameters to inputs, add gradient accumulation ...).

The output of the offline step will be 3 files:
1. Training onnx graph (to be consumed by `Module`).
2. Optimizer onnx graph (to be consumed by `Optimizer`).
3. Checkpoint file that contains all the trainable as well as non-trainable parameters (in a format that is defined by this PR).

For part 3, the offline step contains information on all model parameters (as tensor protos) and requires grad information (separate from the onnx file). The training onnx file does not have the parameters anymore (since they are moved to the inputs). So the `save_checkpoint` api needs to be able to take only the parameter inputs (and the metadata) to save the checkpoint. We can pass the model, but the model will not contain enough information to extract the checkpoint.

Reply:
OK, then how the 3 file is generated? What I am thinking is it is generated by calling save_checkpoint passing the inference model path, along with a trainable param name list.

Reply:
Ok, there is one thing I missed, you need anyway load the TensorProto, it will be there anyway in Python. Reloading from C++ backend won't bring better perf than the implicit copy of pybind. Let me revert this change back to the older way as you suggested. :)

Reply:
Fixed as we synced offline.